soroban-spec/fixtures/ts/src/invoke.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  isConnected,
+  getPublicKey,
+  signTransaction,
+} from "@stellar/freighter-api";
+import * as SorobanClient from 'soroban-client'
+import type { Account, Memo, MemoType, Operation, Transaction } from 'soroban-client';
+import { NETWORK_PASSPHRASE, CONTRACT_ID } from './constants'
+import { Server } from './server'
+
+export type Tx = Transaction<Memo<MemoType>, Operation[]>
+
+export type Simulation = NonNullable<SorobanClient.SorobanRpc.SimulateTransactionResponse['results']>[0]
+
+export type TxResponse = SorobanClient.SorobanRpc.GetTransactionResponse;
+
+export type InvokeArgs = {
+  method: string
+  args?: any[]
+  signAndSend?: boolean
+  fee?: number
+}
+
+/**
+ * Get account details from the Soroban network for the publicKey currently
+ * selected in Freighter. If not connected to Freighter, throws errors. Will
+ * pop up Freighter's modal to request user permissions, if this hasn't been
+ * done already.
+ */
+export async function getAccount(): Promise<Account> {
+  if (!await isConnected()) {
+    throw new Error('Freighter not connected')
+  }
+  const publicKey = await getPublicKey()
+  if (!publicKey) {
+    throw new Error('Freighter not initialized')
+  }
+  return await Server.getAccount(publicKey)
+}
+
+export class NotImplementedError extends Error { }
+
+/**
+ * Invoke a method on the contract-data-example contract.
+ *
+ * Uses Freighter to determine the current user and if necessary sign the transaction.
+ *
+ * @param {string} obj.method - The method to invoke.
+ * @param {any[]} obj.args - The arguments to pass to the method.
+ * @param {boolean} obj.signAndSend - Whether to sign and send the transaction, or just simulate it. Unless the method requires authentication.
+ * @param {number} obj.fee - The fee to pay for the transaction.
+ * @returns The transaction response, or the simulation result if signing isn't required.
+ */
+export async function invoke({ method, args = [], fee = 100, signAndSend = false }: InvokeArgs): Promise<(TxResponse & { xdr: string }) | Simulation> {
+  const account = await getAccount()
+
+  const contract = new SorobanClient.Contract(CONTRACT_ID)
+
+  let tx = new SorobanClient.TransactionBuilder(account, {
+    fee: fee.toString(10),
+    networkPassphrase: NETWORK_PASSPHRASE,
+  })
+    .addOperation(contract.call(method, ...args))
+    .setTimeout(SorobanClient.TimeoutInfinite)
+    .build()
+
+  const simulated = await Server.simulateTransaction(tx)
+  const auths = simulated.results?.[0]?.auth
+  // is it possible for `auths` to be present but empty? Probably not, but let's be safe.
+  if (signAndSend) {
```

Comment:
Suggestion: We could clean this up to be a bit more readable, by inverting this:
```suggestion
  if (!signAndSend) {
    const { results } = simulated
    if (!results || results[0] === undefined) {
      if (simulated.error) {
        throw new Error(simulated.error as unknown as string)
      }
      throw new Error(`Invalid response from simulateTransaction:\n{simulated}`)
    }
    return results[0]
  }
  
  const auths = simulates.results?.[0]?.auth
  // .. The rest of signing & sending
```